protocol/x/affiliates/keeper/keeper_test.go

```diff
@@ -14,3 +20,193 @@ func TestLogger(t *testing.T) {
+			expectError: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			err := k.RegisterAffiliate(ctx, tc.referee, tc.affiliate)
+			if tc.expectError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				affiliate, exists := k.GetReferredBy(ctx, tc.referee)
+				require.True(t, exists)
+				require.Equal(t, tc.affiliate, affiliate)
+			}
+		})
+	}
+}
+
+func TestAddReferredVolume(t *testing.T) {
+	tApp := testapp.NewTestAppBuilder(t).Build()
+	ctx := tApp.InitChain()
+	k := tApp.App.AffiliatesKeeper
+
+	affiliate := "affiliate1"
+	initialVolume := dtypes.NewInt(1000)
+	addedVolume := dtypes.NewInt(500)
+
+	err := k.AddReferredVolume(ctx, affiliate, initialVolume)
+	require.NoError(t, err)
+
+	volume, exists := k.GetReferredVolume(ctx, affiliate)
+	require.True(t, exists)
+	require.Equal(t, initialVolume, volume)
+
+	err = k.AddReferredVolume(ctx, affiliate, addedVolume)
+	require.NoError(t, err)
+
+	updatedVolume, exists := k.GetReferredVolume(ctx, affiliate)
+	require.True(t, exists)
+	require.Equal(t, dtypes.NewInt(1500), updatedVolume)
+}
+
+func TestGetTakerFeeShareViaReferredVolume(t *testing.T) {
+	tApp := testapp.NewTestAppBuilder(t).Build()
+	ctx := tApp.InitChain()
+	k := tApp.App.AffiliatesKeeper
+
+	// Set up affiliate tiers
+	affiliateTiers := types.AffiliateTiers{
+		Tiers: []types.AffiliateTiers_Tier{
+			{Level: 0, ReqReferredVolume: 0, TakerFeeSharePpm: 1000, ReqStakedWholeCoins: 200},
+			{Level: 1, ReqReferredVolume: 1000, TakerFeeSharePpm: 2000, ReqStakedWholeCoins: 1000},
+		},
+	}
+	err := k.UpdateAffiliateTiers(ctx, affiliateTiers)
+	require.NoError(t, err)
+
+	// Register affiliate and referee
+	affiliate := constants.AliceAccAddress.String()
+	referee := constants.BobAccAddress.String()
+	stakingKeeper := tApp.App.StakingKeeper
+	err = stakingKeeper.SetDelegation(ctx,
+		stakingtypes.NewDelegation(constants.AliceAccAddress.String(),
+			constants.AliceValAddress.String(), math.LegacyNewDecFromBigInt(big.NewInt(100))))
+	require.NoError(t, err)
+	err = k.RegisterAffiliate(ctx, referee, affiliate)
+	require.NoError(t, err)
+
+	// Add referred volume for affiliate
+	err = k.AddReferredVolume(ctx, affiliate, dtypes.NewInt(500))
+	require.NoError(t, err)
+
+	// Get taker fee share for referee
+	affiliateAddr, feeSharePpm, exists, err := k.GetTakerFeeShare(ctx, referee)
+	require.NoError(t, err)
+	require.True(t, exists)
+	require.Equal(t, affiliate, affiliateAddr)
+	require.Equal(t, uint32(1000), feeSharePpm)
+
+	// Add more referred volume to upgrade tier
+	err = k.AddReferredVolume(ctx, affiliate, dtypes.NewInt(500))
+	require.NoError(t, err)
+
+	// Get updated taker fee share for referee
+	affiliateAddr, feeSharePpm, exists, err = k.GetTakerFeeShare(ctx, referee)
+	require.NoError(t, err)
+	require.True(t, exists)
+	require.Equal(t, affiliate, affiliateAddr)
+	require.Equal(t, uint32(2000), feeSharePpm)
+}
+
+func TestGetTakerFeeShareViaStakedAmount(t *testing.T) {
```

Comment:
Could we add a test case where `ReferredVolume` qualifies for tier 2, and `StakedAmount` qualifies for tier 3 (so overall tier is 3)?